docs/src/adr/014adr-precise-records.md

```diff
@@ -0,0 +1,668 @@
+          /\ HandleAppendEntriesRequest(i, j, m)
+       \/ /\ m.mtype = AppendEntriesResponse
+          /\ \/ DropStaleResponse(i, j, m)
+             \/ HandleAppendEntriesResponse(i, j, m)
+    ...
+Next ==
+    ...
+           \/ \E m \in DOMAIN messages : Receive(m)
+           \/ \E m \in DOMAIN messages : DuplicateMessage(m)
+           \/ \E m \in DOMAIN messages : DropMessage(m)
+    ...
+```
+
+### 3. Other issues
+
+As can be seen from the Paxos example, we should take care of sets of
+records that are used as sets of variants:
+
+```tla
+Message ==      [type : {"1a"}, bal : Ballot]
+           \cup [type : {"1b"}, acc : Acceptor, bal : Ballot, 
+                 mbal : Ballot \cup {-1}, mval : Value \cup {None}]
+           \cup [type : {"2a"}, bal : Ballot, val : Value]
+           \cup [type : {"2b"}, acc : Acceptor, bal : Ballot, val : Value]
+```           
+
+## 4. Options
+
+There are several solutions to the issue of precise record typing.
+
+### 4.1. Support only plain records
+
+In this case, we would only allow to mix records that have exactly the same
+shape. As a result, when we use variants, we would have to add extra
+fields to all of them. This solution is not very different from the current
+implementation of record type checking, though it would allow us to quickly
+detect spelling errors.
+
+**Blocker.** This does not look like a real solution, as it would immediately
+render the existing examples invalid. Moreover, they would be no obvious way to
+repair these examples.
+
+### 4.2. Support plain records and variants, but no row typing
+
+In this scenario, the type checker would issue an error, if a record expression
+accesses a field that is outside of its type:
+
+```tla
+FieldAccess ==
+  LET m == [ a |-> 2, b |-> "B" ] IN
+  /\ m.a > 1        \* type OK
+  /\ m.b = "B"      \* type OK
+  /\ m.c = { 1, 2 } \* should flag a type error
+```
+
+Moreover, the following example would require a type annotation:
+
+```tla
+RowAccess(m) ==
+    m.a > 0         \* should flag a type error
+```
+
+In the above example, the type checker would not be able to infer the type of
+`m` and would require an explicit type annotation:
+
+
+```tla
+\* @type: [ a: Int, b: Str ];
+RowAccess(m) ==
+    m.a > 0         \* should not flag a type error
+```
+
+**Blocker.** It is not clear to me, what type we would assign to the record
+access operator. Although the type of `m` in `FieldAccess` is obvious to a
+human reader, as it simply requires us to do the top-to-the-bottom type
+propagation, Snowcat constructs a set of equality constraints that are solved
+by unification. This approach requires that we capture the record access
+operator `.` as an equality over type variables and types.
+
+### 4.3. Support plain records and variants, including row typing
+
+This solution follows the approach outlined by [Leijen05][], except that we
+require field names to be unique, not scoped. We present our approach in the
+next section.
```

Comment:
One point of concern, discussed briefly in our standup, but which I should flag here. In `Leijen05`, the author notes that 

> Row extension. Wand [35, 36] was the first to introduce row vari-
> ables, and described a type system that is very close to our proposal.
> The types of the basic operations are given as type rules, but these
> rules have basically the same effect as the type signatures in our
> system. However, this system also uses an ambiguous interpreta-
> tion of free extension as a mixture of update and extension. This
> gives rise to fundamental problems in the type system. Take for
> example the following expression:
>
> \r → if True then {x = 1 | r } else {x = 2 | {}}
>
> If extension overwrites previous fields, we can assign two different
> types to r , namely {x :: Int} or the empty record {}. This is
> why Wand’s original record system proved incomplete [35]. With
> scoped labels, the choice is unambiguous and r must have type {}.
> We show completeness (and soundness) of type inference formally
> in Section 7.

IIUC, the idea is that scoped labels are necessary to differentiate the extension and update operations, because otherwise any extension is necessarily an update in the event that the extended record already has a field with the same name. As a result, in the example given, it is impossible to give a unique type to `r`.  

For example, suppose `r = {x = 3}`. If extension entails update, then `{x = 1 | r} = {x = 1}`. This type checks. But now suppose `r = {}`, then, again `{x = 1 | r} = {x = 1}`. So we end up with ambiguous typing for `r`. This problem is avoided by scoped labels, because when `r = {x = 3}`, then `{x = 1 | r } = {(|x = 3::(|x=1::(||)|)|)}, but when `r = {}`, then we just get `{x = 1 | r} = {x = 1}`. This prevents us from ending up with an undecidable typing.

Given this warning in the paper, if my concern here seems sensible, then it is probably worth devoting a section to exploring how we can ensure that our decision to not adopt scoped labels won't leave us with problems. I would imagine there's more recent work that address this problem?

(Note that I tried to find the technical report to which the other refers readers to look for the proof of completeness, in the hopes that we might see how exactly the scoped labels are involved, but I can't find it anywhere online, except in the citation of paper.)

Reply:
But we do not have record extension in TLA+. I don't even think we would allow `EXCEPT` to change types of the fields. So yes, probably the type system of Wand [35, 36] is incomplete. But I do not think this applies to us. The problem with the recent research is that people are pushing more and more features into papers, because they cannot publish otherwise. We actually need a relatively simple type system and relevant exposition can be only found in the old papers.

Reply:
Ah ok. Right if we have no record extension, then that would address this concern.

Reply:
Quick note here that @gabrielamafra helped me understand this point bout TLA+ not supporting record extension. I had assumed that `EXCEPT` could be used to add new fields to records, but apparently, in both TLC and Apalache, using EXCEPT with a record field which is not already in the record leads to the update just be ignored? (Gabriela noted that this also issues a warning in TLC).